### ISSUE TITLE
Sortmerna 4.2 fails to process concatenated gz files

### ISSUE THREAD

**Issue by esauta** [biocore/sortmerna#272]
Hi,
I'm using the latest version of sortmerna (4.2). I want to filter out rRNA reads from paired reads fastq.gz samples. 
After checking idx folder with indexed rRNA fasta files, sortmerna stops at this point without any progress: 

`[Index:117] TODO: a better validation using an index descriptor to decide on indexing`

Moreover, it did not create the "out" directory; there are only idx and kvdb directories. Could you help me?
Thank you

Here is my command:
```
./bin/sortmerna -ref /data/rRNA_databases/silva-bac-16s-id90.fasta -ref /data/rRNA_databases/silva-euk-28s-id98.fasta -ref /data/rRNA_databases/rfam-5.8s-database-id98.fasta -ref /data/rRNA_databases/rfam-5s-database-id98.fasta -ref /data/rRNA_databases/silva-arc-16s-id95.fasta -ref /data/rRNA_databases/silva-bac-23s-id98.fasta -ref /data/rRNA_databases/silva-arc-23s-id98.fasta -ref /data/rRNA_databases/silva-euk-18s-id95.fasta -reads /input_data/sample_S2_R1_001.fastq.gz -reads /input_data/sample_S2_R2_001.fastq.gz -workdir ./sortmerna/Run -fastx -other sample_filtered.fastq.gz -aligned sample_aligned_rRNA.fastq.gz  --paired_in -v --threads 5

And the related trace:

[process:1369] === Options processing starts ... ===

Found value: ./bin/sortmerna
Found flag: -ref
Found value: /data/rRNA_databases/silva-bac-16s-id90.fasta of previous flag: -ref
Found flag: -ref
Found value: /data/rRNA_databases/silva-euk-28s-id98.fasta of previous flag: -ref
Found flag: -ref
Found value: /data/rRNA_databases/rfam-5.8s-database-id98.fasta of previous flag: -ref
Found flag: -ref
Found value: /data/rRNA_databases/rfam-5s-database-id98.fasta of previous flag: -ref
Found flag: -ref
Found value: /data/rRNA_databases/silva-arc-16s-id95.fasta of previous flag: -ref
Found flag: -ref
Found value: /data/rRNA_databases/silva-bac-23s-id98.fasta of previous flag: -ref
Found flag: -ref
Found value: /data/rRNA_databases/silva-arc-23s-id98.fasta of previous flag: -ref
Found flag: -ref
Found value: /data/rRNA_databases/silva-euk-18s-id95.fasta of previous flag: -ref
Found flag: -reads
Found value: /input_data/sample_S2_R1_001.fastq.gz of previous flag: -reads
Found flag: -reads
Found value: /input_data/sample_S2_R2_001.fastq.gz of previous flag: -reads
Found flag: -workdir
Found value:  ./sortmerna/Run of previous flag: -workdir
Found flag: -fastx
Previous flag: -fastx is Boolean. Setting to True
Found flag: -other
Found value: sample_filtered.fastq.gz of previous flag: -other
Found flag: -aligned
Found value: sample_aligned_rRNA.fastq.gz of previous flag: -aligned
Found flag: --paired_in
Previous flag: --paired_in is Boolean. Setting to True
Found flag: -v
Previous flag: -v is Boolean. Setting to True
Found flag: --threads
Found value: 5 of previous flag: --threads
[opt_workdir:1066] Using WORKDIR: [" ./sortmerna/Run/"] as specified
[process:1453] Processing option: aligned with value: sample_aligned_rRNA.fastq.gz
[process:1453] Processing option: fastx with value: 
[process:1453] Processing option: other with value: sample_filtered.fastq.gz
[process:1453] Processing option: paired_in with value: 
[process:1453] Processing option: reads with value: /input_data/sample_S2_R1_001.fastq.gz
[opt_reads:73] Processing reads file [1] out of total [2] files
[process:1453] Processing option: reads with value: /input_data/sample_S2_R2_001.fastq.gz
[opt_reads:73] Processing reads file [2] out of total [2] files
[process:1453] Processing option: ref with value: ./data/rRNA_databases/silva-bac-16s-id90.fasta
[opt_ref:166] Processing reference [1] out of total [8] references
[opt_ref:220] File ["./sortmerna/data/rRNA_databases/silva-bac-16s-id90.fasta"] exists and is readable
[process:1453] Processing option: ref with value: /data/rRNA_databases/silva-euk-28s-id98.fasta
[opt_ref:166] Processing reference [2] out of total [8] references
[opt_ref:220] File ["./sortmerna/data/rRNA_databases/silva-euk-28s-id98.fasta"] exists and is readable
[process:1453] Processing option: ref with value: /data/rRNA_databases/rfam-5.8s-database-id98.fasta
[opt_ref:166] Processing reference [3] out of total [8] references
[opt_ref:220] File ["./sortmerna/data/rRNA_databases/rfam-5.8s-database-id98.fasta"] exists and is readable
[process:1453] Processing option: ref with value: /data/rRNA_databases/rfam-5s-database-id98.fasta
[opt_ref:166] Processing reference [4] out of total [8] references
[opt_ref:220] File ["./sortmerna/data/rRNA_databases/rfam-5s-database-id98.fasta"] exists and is readable
[process:1453] Processing option: ref with value: /data/rRNA_databases/silva-arc-16s-id95.fasta
[opt_ref:166] Processing reference [5] out of total [8] references
[opt_ref:220] File ["./sortmerna/data/rRNA_databases/silva-arc-16s-id95.fasta"] exists and is readable
[process:1453] Processing option: ref with value: ./data/rRNA_databases/silva-bac-23s-id98.fasta
[opt_ref:166] Processing reference [6] out of total [8] references
[opt_ref:220] File ["./sortmerna/data/rRNA_databases/silva-bac-23s-id98.fasta"] exists and is readable
[process:1453] Processing option: ref with value: /data/rRNA_databases/silva-arc-23s-id98.fasta
[opt_ref:166] Processing reference [7] out of total [8] references
[opt_ref:220] File ["./sortmerna/data/rRNA_databases/silva-arc-23s-id98.fasta"] exists and is readable
[process:1453] Processing option: ref with value: /data/rRNA_databases/silva-euk-18s-id95.fasta
[opt_ref:166] Processing reference [8] out of total [8] references
[opt_ref:220] File ["./sortmerna/data/rRNA_databases/silva-euk-18s-id95.fasta"] exists and is readable
[process:1453] Processing option: threads with value: 5
[process:1453] Processing option: v with value: 

[process:1473] === Options processing done ===
[validate_kvdbdir:1252] Key-value DB location "/sortmerna/Run/kvdb"
[validate_kvdbdir:1288] Creating KVDB directory: "/sortmerna/Run/kvdb"
[validate_idxdir:1233] IDX directory: "/sortmerna/Run/idx" exists and is not empty

  Program:      SortMeRNA version 4.2.0

[main:63] Running command:
./bin/sortmerna -ref /data/rRNA_databases/silva-bac-16s-id90.fasta -ref /data/rRNA_databases/silva-euk-28s-id98.fasta -ref /data/rRNA_databases/rfam-5.8s-database-id98.fasta -ref /data/rRNA_databases/rfam-5s-database-id98.fasta -ref /data/rRNA_databases/silva-arc-16s-id95.fasta -ref /data/rRNA_databases/silva-bac-23s-id98.fasta -ref /data/rRNA_databases/silva-arc-23s-id98.fasta -ref /data/rRNA_databases/silva-euk-18s-id95.fasta -reads /input_data/sample_S2_R1_001.fastq.gz -reads /input_data/sample_S2_R2_001.fastq.gz -workdir ./sortmerna/Run -fastx -other sample_filtered.fastq.gz -aligned sample_aligned_rRNA.fastq.gz  --paired_in -v --threads 5 
[Index:95] Index file ["/sortmerna/Run/idx/15734375058464002811.bursttrie_0.dat"] already exists and is not empty.
[Index:95] Index file ["/sortmerna/Run/idx/15734375058464002811.pos_0.dat"] already exists and is not empty.
[Index:95] Index file ["/sortmerna/Run/idx/15734375058464002811.kmer_0.dat"] already exists and is not empty.
[Index:95] Index file ["/sortmerna/Run/idx/15734375058464002811.stats"] already exists and is not empty.
[Index:95] Index file ["/sortmerna/Run/idx/1845323523482939374.bursttrie_0.dat"] already exists and is not empty.
[Index:95] Index file ["/sortmerna/Run/idx/1845323523482939374.pos_0.dat"] already exists and is not empty.
[Index:95] Index file ["/sortmerna/Run/idx/1845323523482939374.kmer_0.dat"] already exists and is not empty.
[Index:95] Index file ["/sortmerna/Run/idx/1845323523482939374.stats"] already exists and is not empty.
[Index:95] Index file ["/sortmerna/Run/idx/2169995244134016533.bursttrie_0.dat"] already exists and is not empty.
[Index:95] Index file ["/sortmerna/Run/idx/2169995244134016533.pos_0.dat"] already exists and is not empty.
[Index:95] Index file ["/sortmerna/Run/idx/2169995244134016533.kmer_0.dat"] already exists and is not empty.
[Index:95] Index file ["/sortmerna/Run/idx/2169995244134016533.stats"] already exists and is not empty.
[Index:95] Index file ["/sortmerna/Run/idx/13019673092862722585.bursttrie_0.dat"] already exists and is not empty.
[Index:95] Index file ["/sortmerna/Run/idx/13019673092862722585.pos_0.dat"] already exists and is not empty.
[Index:95] Index file ["/sortmerna/Run/idx/13019673092862722585.kmer_0.dat"] already exists and is not empty.
[Index:95] Index file ["/sortmerna/Run/idx/13019673092862722585.stats"] already exists and is not empty.
[Index:95] Index file ["/sortmerna/Run/idx/3436099190853847617.bursttrie_0.dat"] already exists and is not empty.
[Index:95] Index file ["/sortmerna/Run/idx/3436099190853847617.pos_0.dat"] already exists and is not empty.
[Index:95] Index file ["/sortmerna/Run/idx/3436099190853847617.kmer_0.dat"] already exists and is not empty.
[Index:95] Index file ["/sortmerna/Run/idx/3436099190853847617.stats"] already exists and is not empty.
[Index:95] Index file ["/sortmerna/Run/idx/17299952793705614139.bursttrie_0.dat"] already exists and is not empty.
[Index:95] Index file ["/sortmerna/Run/idx/17299952793705614139.pos_0.dat"] already exists and is not empty.
[Index:95] Index file ["/sortmerna/Run/idx/17299952793705614139.kmer_0.dat"] already exists and is not empty.
[Index:95] Index file ["/sortmerna/Run/idx/17299952793705614139.stats"] already exists and is not empty.
[Index:95] Index file ["/sortmerna/Run/idx/3400685301612210653.bursttrie_0.dat"] already exists and is not empty.
[Index:95] Index file ["/sortmerna/Run/idx/3400685301612210653.pos_0.dat"] already exists and is not empty.
[Index:95] Index file ["/sortmerna/Run/idx/3400685301612210653.kmer_0.dat"] already exists and is not empty.
[Index:95] Index file ["/sortmerna/Run/idx/3400685301612210653.stats"] already exists and is not empty.
[Index:95] Index file ["/sortmerna/Run/idx/2700646386527218729.bursttrie_0.dat"] already exists and is not empty.
[Index:95] Index file ["/sortmerna/Run/idx/2700646386527218729.pos_0.dat"] already exists and is not empty.
[Index:95] Index file ["/sortmerna/Run/idx/2700646386527218729.kmer_0.dat"] already exists and is not empty.
[Index:95] Index file ["/sortmerna/Run/idx/2700646386527218729.stats"] already exists and is not empty.
[Index:116] Found 32 non-empty index files. Skipping indexing.
[Index:117] TODO: a better validation using an index descriptor to decide on indexing
```

**Comment by biocodz** [comment]
could you, please, try
```
-other sample_filtered
-aligned sample_aligned_rRNA
```
if  no luck then, please, try without 'other' and 'aligned' at all i.e. leave it to default values just to see how it goes.

**Comment by esauta** [comment]
Using these lines: 

-other sample_filtered
-aligned sample_aligned_rRNA

Nothing changed; sortmerna is blocked at the same point (at [Index:117] TODO: a better validation using an index descriptor to decide on indexing) without creating the "out" folder.

When I removed also those lines, sortmerna creates the "out" folder, but it is still blocked at:
 [Index:117] TODO: a better validation using an index descriptor to decide on indexing

And the log file still remain empty

**Comment by biocodz** [comment]
it chokes right at the start of processing the reads files.
What's the output of
```
gzip -l /input_data/sample_S2_R1_001.fastq.gz
gzip -l /input_data/sample_S2_R2_001.fastq.gz
```

**Comment by esauta** [comment]
gzip -l /input_data/sample_S2_R1_001.fastq.gz
compressed        uncompressed     ratio uncompressed_name
3989034190           30287                               -13170680.1%

gzip -l /input_data/sample_S2_R2_001.fastq.gz
compressed        uncompressed     ratio uncompressed_name
3995132823              30908                          -12925785.8%

**Comment by biocodz** [comment]
looks abnormal.
Could you try re-compressing:
```
gzip -cd /input_data/sample_S2_R1_001.fastq.gz | gzip > new1.fastq.gz
gzip -cd /input_data/sample_S2_R2_001.fastq.gz | gzip > new2.fastq.gz
gzip -l new1.fastq.gz
gzip -l new2.fastq.gz
```

**Comment by esauta** [comment]
These files are the results of a whole genome sequencing, after a standard demultiplexing step, so the huge size is normal. 
I try your cmd on a single file, and this is the output:
gzip -l new2.fastq.gz

compressed        uncompressed  ratio uncompressed_name
2888926802          1143736465      -152.6%

**Comment by biocodz** [comment]
I'm not sure whether it's the size or something in the file structure that confuses the program. Unfortunately, each time I see gzip failing to calculate the compression ratio, sortmerna fails (it uses zlib, like gzip). 
I didn't encounter this problem with files of a few gig size, but may be at some size it would fail. I'll have to work on this.

**Comment by morgansobol** [comment]
Hi @biocodz, 

I believe I am having the same issue, however, trying to use an unzipped file does not work for me. 
Here is my bash script `#!/bin/bash`
#SBATCH --job-name=Triple
#SBATCH --time=48:00:00
#SBATCH --mem=50gb
#SBATCH --output=Triple_4.2.out
./sortmerna --ref ../data/rRNA_databases/silva-bac-23s-id98.fasta --ref ../data/rRNA_databases/silva-bac-16s-id90.fasta --reads /pfs/work7/workspace/scratch/fh9393-ahmed_transcriptome-0/new_workflow/results/2_trimmed_output/Triple_S7_R1_001_val_1.fq --reads /pfs/work7/workspace/scratch/fh9393-ahmed_transcriptome-0/new_workflow/results/2_trimmed_output/Triple_S7_R2_001_val_2.fq --workdir /home/kit/ibg/fh9393/sortmerna/run_4.2/ --aligned --other --fastx`

Exection log attached 
[Triple_4.2.txt](https://github.com/biocore/sortmerna/files/5659326/Triple_4.2.txt)

As you can see, it gets stuck at "[init:218] Testing file: "/home/kit/ibg/fh9393/sortmerna/run_4.2/out/aligned.log""

Files sizes:
`-rw-r--r-- 1 fh9393 ibg 9134497279 Dec  6 15:19 Triple_S7_R1_001_val_1.fq
-rw-r--r-- 1 fh9393 ibg 9180858013 Dec  6 15:19 Triple_S7_R2_001_val_2.fq`

The output from the top command shows it is not running (I think) 
![Screen Shot 2020-12-08 at 12 24 56 PM](https://user-images.githubusercontent.com/36306082/101485576-a819c380-395b-11eb-8b2f-316b3342f50e.png)

My fastq files are outputs from TrimGalore. I also tried using Sortmerna version 4.0, and there it got stuck at "[store_to_db:421] Stored Reads statistics to DB:"

Any ideas?

Thanks!
Morgan

**Comment by biocodz** [comment]
@morgansobol 
hi,
the alignment in your case hasn't even started. Although there were no problems reading the reads files, it seems there was a problem processing the refs. The trace indicates the program could find and read the refs, yet something prevented the processing. I've seen similar problems when there were some extraneous chars like new line in the command line. 
When the program runs it's obvious from the progressing trace. You can see a sample [here](https://github.com/biocore/sortmerna/wiki/Sample-execution-trace)

**Comment by Xiaoyu2425** [comment]
Hi, I have exactly the same issue with @esauta - failure after [Index:117] with .fastq.gz files - but working with .fastq files. I guess something regarding .gz files processing should be adjusted in this program.

**Comment by biocodz** [comment]
we are close to a new release that introduces enhanced processing of archived files including automatic recognition and validation no matter the extension, and producing gzipped output. The release fixes many problems and provides a better diagnostics to make easier to tackle problems like the current one.

**Comment by Xiaoyu2425** [comment]
> we are close to the next release that introduces enhanced processing of archived files including automatic recognition and validation no matter the extension, and producing gzipped output. The release fixes many problems and provides a better diagnostics to make easier to tackle problems like the current one.

Really appreciate it! Happy new year.

**Comment by cflsjt** [comment]
Hi, I am having the same issues with this. I have done some tests with a small subset but using bgzip for compression - no output. Following a previous suggestion to recompress with gzip it does work. 

We received TB of fastq.gz data from our providers which are not compatible - hopefully a bug fix is coming along?

**Comment by ekopylova** [comment]
Hello @cflsjt ! Can you please test the latest [release 4.3.1](https://github.com/biocore/sortmerna/releases/tag/v4.3.1) and see if the error persists? Thanks!!

**Comment by cflsjt** [comment]
> Hello @cflsjt ! Can you please test the latest [release 4.3.1](https://github.com/biocore/sortmerna/releases/tag/v4.3.1) and see if the error persists? Thanks!!

Apologies this is the latest version that I used v4.3.1

**Comment by cflsjt** [comment]
@biocodz my colleague has looked through some of the unit tests and this needs sorted there too. We currently use htslib for bgzip compression but our sequence files form providers are all bgzip. 
I can create a new issue for v4.3.1?

**Comment by biocodz** [comment]
`bgzip` would be a new feature. Sure, file an issue for this.

**Comment by biocodz** [comment]
supposedly fixed in [v4.3.3](https://github.com/biocore/sortmerna/releases/tag/v4.3.3)
